plugin/pkg/admission/wmftoolsenforcer/admission.go

```diff
@@ -0,0 +1,142 @@
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+func init() {
+	admission.RegisterPlugin("UidEnforcer", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		return NewUidEnforcer(client), nil
+	})
+}
+
+// plugin contains the client used by the uidenforcer admin controller
+type plugin struct {
+	*admission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+// NewSecurityContextDeny creates a new instance of the SecurityContextDeny admission controller
+func NewUidEnforcer(client client.Interface) admission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return client.Namespaces().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Namespaces().Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Namespace{},
+		store,
+		5*time.Minute,
+	)
+	reflector.Run()
+	return &plugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		client:  client,
+		store:   store,
+	}
+}
+
+// This will verify the following:
+//  - Not a namespace modification from a namespaced user for their own namespace
+//  - User object has a numeric uid
+//  - Namespace object has an annotation called RunAsUser that's an integer
+//
+// If after all this there's no SecurityContext on each Container with a RunAsUser set to the same RunAsUser, it'll be set
+func (p *plugin) Admit(a admission.Attributes) (err error) {
+	if a.GetResource() != string(api.ResourcePods) && a.GetResource() != "namespace" {
+		return nil
+	}
+
+	// HACK: This should probably be a separate AdmissionController
+	hackNamespace, ok := a.GetObject().(*api.Namespace)
+	if ok {
+		user := a.GetUserInfo()
+		if user == nil {
+			return apierrors.NewBadRequest("uidenforcer admission controller can not be used if there is no user set")
+		}
+
+		// This is a namespace!
+		// We hope only users can directly change this!
+		// So namespaced users can not affect other namespaces, but can change the
+		// annotations on their own namespace. let's deny that
+		if hackNamespace.Name == user.GetName() {
+			return apierrors.NewBadRequest("Can not modify users' own namespace")
+		}
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	namespaceObj, exists, err := p.store.Get(&api.Namespace{
+		ObjectMeta: api.ObjectMeta{
+			Name:      a.GetNamespace(),
+			Namespace: "",
+		},
+	})
+
+	if !exists {
+		return apierrors.NewBadRequest("Namespace " + a.GetNamespace() + " not found")
+	}
+
+	if err != nil {
+		return apierrors.NewBadRequest("Everything must be in a namespace!")
+	}
+	namespace := namespaceObj.(*api.Namespace)
+
+	if _, ok := namespace.Annotations["RunAsUser"]; !ok {
```

Comment:
So right now I've nothing in the kube-system namespace (no working skydns yet), and you're right I'll have to figure something out for that. However, now that I've whitelisted the list of allowed resources even more maybe this entire thing can go.